### PR TITLE
treewide: abort -> throw where it makes sense

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
I don't really care about this getting merged, but I'd like to find out whether there's any reason for using aborts. Please let me know if you have any valid use case, because I couldn't find any. I also asked many people but nobody had one.

I have *NOT* tested this in any way, let me know if this breaks anything, which might just be a reason for having aborts then.

Note that I'm looking for *actual* use cases, not theoretical ones.

Ping @Profpatsch

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ericson2314** [comment]
Yeah I 've never seen any consistency in one or the other. I assume people use one cause they didn't know of the other.

File: default.nix [hunk 1]
```diff
@@ -2,7 +2,7 @@ let requiredVersion = import ./lib/minver.nix; in
 
 if ! builtins ? nixVersion || builtins.compareVersions requiredVersion builtins.nixVersion == 1 then
 
-  abort ''
+  throw ''
```
Comment (Mic92) [inline default.nix]: Is this maybe a special case? Some commands like `nix-env -qa` will silently ignore `throw`.

**Comment by Profpatsch** [comment]
As @Mic92 said, `abort` can’t be skipped, while `throw` is skipped by `nix-env`. Most changes from `lib` signal programmer errors, so should stay `abort`. It doesn’t make sense for packages to use `abort` however.

File: lib/attrsets.nix [hunk 2]
```diff
@@ -72,7 +72,7 @@ rec {
   */
   getAttrFromPath = attrPath: set:
     let errorMsg = "cannot find attribute `" + concatStringsSep "." attrPath + "'";
-    in attrByPath attrPath (abort errorMsg) set;
+    in attrByPath attrPath (throw errorMsg) set;
```
Comment (Profpatsch) [inline lib/attrsets.nix]: +1

File: lib/generators.nix [hunk 3]
```diff
@@ -31,7 +31,7 @@ rec {
    * suitable for bash scripts but not much else.
    */
   mkValueStringDefault = {}: v: with builtins;
-    let err = t: v: abort
+    let err = t: v: throw
```
Comment (Profpatsch) [inline lib/generators.nix]: `abort`, is a programmer error.

File: lib/generators.nix [hunk 4]
```diff
@@ -171,7 +171,7 @@ rec {
                        fna);
       in if fna == {}    then "<λ>"
                          else "<λ:{${showFnas}}>"
-    else abort "generators.toPretty: should never happen (v = ${v})";
+    else throw "generators.toPretty: should never happen (v = ${v})";
```
Comment (Profpatsch) [inline lib/generators.nix]: `abort`, is a failed pattern match and thus a programmer error.

File: lib/generators.nix [hunk 5]
```diff
@@ -184,7 +184,7 @@ rec {
       if isList x   then list ind x else
       if isAttrs x  then attrs ind x else
       if isFloat x  then float ind x else
-      abort "generators.toPlist: should never happen (v = ${v})";
+      throw "generators.toPlist: should never happen (v = ${v})";
```
Comment (Profpatsch) [inline lib/generators.nix]: as above

File: lib/modules.nix [hunk 6]
```diff
@@ -671,7 +671,7 @@ rec {
     let
       fromOpt = getAttrFromPath from options;
       toOf = attrByPath to
-        (abort "Renaming error: option `${showOption to}' does not exist.");
+        (throw "Renaming error: option `${showOption to}' does not exist.");
```
Comment (Profpatsch) [inline lib/modules.nix]: Also a programmer error, `abort`.

File: lib/options.nix [hunk 7]
```diff
@@ -60,14 +60,14 @@ rec {
     else throw "Cannot merge definitions of `${showOption loc}' given in ${showFiles (getFiles defs)}.";
 
   mergeOneOption = loc: defs:
-    if defs == [] then abort "This case should never happen."
+    if defs == [] then throw "This case should never happen."
```
Comment (Profpatsch) [inline lib/options.nix]: `abort`

File: lib/options.nix [hunk 8]
```diff
@@ -60,14 +60,14 @@ rec {
...
     else if length defs != 1 then
       throw "The unique option `${showOption loc}' is defined multiple times, in ${showFiles (getFiles defs)}."
     else (head defs).value;
 
   /* "Merge" option definitions by checking that they all have the same value. */
   mergeEqualOption = loc: defs:
-    if defs == [] then abort "This case should never happen."
+    if defs == [] then throw "This case should never happen."
```
Comment (Profpatsch) [inline lib/options.nix]: `abort`

File: nixos/lib/testing.nix [hunk 9]
```diff
@@ -91,7 +91,7 @@ in rec {
 
       testDriverName = with builtins;
         if testNameLen > maxTestNameLen then
-          abort ("The name of the test '${name}' must not be longer than ${toString maxTestNameLen} " +
+          throw ("The name of the test '${name}' must not be longer than ${toString maxTestNameLen} " +
```
Comment (Profpatsch) [inline nixos/lib/testing.nix]: Programmer error, `abort`.

**Review by Profpatsch — CHANGES_REQUESTED** [review]
I commented on the first few changes, it looks to me like nearly all of these are programmer errors that should abort the evaluation whenever they happen (not skippable, like with `throw`).
So I’m against this change.

**Comment by FRidh** [comment]
Added `"needs: documentation"` label because clearly that's the issue here.

**Comment by Infinisil** [comment]
Can this be merged like this? This now only changes the `abort`s that should've been `throw`s from the beginning.

**Comment by Mic92** [comment]
I am not sure why @edolstra was against the pull request and if those issues are still present.

**Comment by Infinisil** [comment]
Can't be bothered with this

**Pull request closed without merging**